File: internal/stats/grade.go
// grade.go computes the letter grade shown on the stats-card-with-grade widget
// (gaka-hsj). It is a faithful port of github-readme-stats' rank algorithm
// (src/calculateRank.js): the same two CDFs, the same weighted-blend shape, the
// same percentile thresholds and level ladder. The ONLY local adaptation is the
// metric set — GitHub's commits/prs/issues/reviews/stars/followers are swapped
// for coding-time equivalents, each with a tunable median. Lower percentile is
// better, exactly like upstream (S ≈ top 1%).
package stats

import (
	"math"

	"github.com/TheBranchDriftCatalyst/boomtime/internal/model"
)

// gradeLevels and gradeThresholds are verbatim from github-readme-stats:
// the level at the first threshold with percentile <= t.
var (
	gradeLevels     = []string{"S", "A+", "A", "A-", "B+", "B", "B-", "C+", "C"}
	gradeThresholds = []float64{1, 12.5, 25, 37.5, 50, 62.5, 75, 87.5, 100}
)

// exponentialCDF is upstream's exponential_cdf: 1 - 2^-x.
func exponentialCDF(x float64) float64 { return 1 - math.Pow(2, -x) }

// logNormalCDF is upstream's log_normal_cdf: x / (1 + x).
func logNormalCDF(x float64) float64 { return x / (1 + x) }

// GradeConfig holds the medians + weights for each metric. Weights mirror the
// upstream slot each metric replaces (commits=2, prs=3, issues=1, reviews=1,
// stars=4, followers=1). Medians are the local tuning knobs — "a median coder
// scores 50th percentile" is the calibration intent.
type GradeConfig struct {
	StreakMedian float64 // days; upstream commits slot (weight 2, exponential)
	StreakWeight float64

	ActiveMedian float64 // active-day ratio x100; upstream prs slot (weight 3, exponential)
	ActiveWeight float64

	LanguagesMedian float64 // distinct languages; upstream issues slot (weight 1, exponential)
	LanguagesWeight float64

	ProjectsMedian float64 // distinct projects; upstream reviews slot (weight 1, exponential)
	ProjectsWeight float64

	DailyAvgMedian float64 // seconds/day; upstream stars slot (weight 4, log-normal)
	DailyAvgWeight float64

	HoursMedian float64 // total hours in range; upstream followers slot (weight 1, log-normal)
	HoursWeight float64

	// MinRangeDays floors the active-ratio denominator so a 1-day range can't
	// score 100% consistency (short-range inflation guard).
	MinRangeDays int
}

// DefaultGradeConfig is the shipped calibration (see gaka-hsj plan). Kept as a
// package-level var so cmd/boomtime (via ApplyGradeConfigFromEnv) can tune it
// once at boot without threading a config value through every renderer.
var DefaultGradeConfig = GradeConfig{
	StreakMedian: 5, StreakWeight: 2,
	ActiveMedian: 50, ActiveWeight: 3,
	LanguagesMedian: 3, LanguagesWeight: 1,
	ProjectsMedian: 3, ProjectsWeight: 1,
	DailyAvgMedian: 7200, DailyAvgWeight: 4,
	HoursMedian: 40, HoursWeight: 1,
	MinRangeDays: 7,
}

// SubScore is one metric's contribution — kept on the result for debugging and
// future tuning (a widget tooltip or a /grade debug endpoint can surface it).
type SubScore struct {
	Metric string  `json:"metric"`
	Raw    float64 `json:"raw"`    // metric value before normalization
	Median float64 `json:"median"` // config median it was divided by
	Score  float64 `json:"score"`  // CDF output, 0..1
	Weight float64 `json:"weight"`
}

// GradeResult is the computed grade for one StatsPayload.
type GradeResult struct {
	Level      string     `json:"level"`      // S, A+, A, A-, B+, B, B-, C+, C
	Percentile float64    `json:"percentile"` // 0..100, LOWER is better (upstream convention)
	Subs       []SubScore `json:"subs"`
}

// Grade computes the letter grade for a stats payload with the default config.
func Grade(p *model.StatsPayload) GradeResult { return GradeWith(p, DefaultGradeConfig) }

// GradeWith is Grade with an explicit config — pure, payload-only, no DB.
func GradeWith(p *model.StatsPayload, cfg GradeConfig) GradeResult {
	rangeDays := len(p.DailyTotal)
	activeDays := 0
	for _, s := range p.DailyTotal {
		if s > 0 {
			activeDays++
		}
	}
	denom := max(rangeDays, cfg.MinRangeDays)
	activeRatio := float64(activeDays) / float64(denom) * 100

	subs := []SubScore{
		{Metric: "streak", Raw: float64(longestStreak(p.DailyTotal)), Median: cfg.StreakMedian, Weight: cfg.StreakWeight},
		{Metric: "activeDays", Raw: activeRatio, Median: cfg.ActiveMedian, Weight: cfg.ActiveWeight},
		{Metric: "languages", Raw: float64(p.LanguagesCount), Median: cfg.LanguagesMedian, Weight: cfg.LanguagesWeight},
		{Metric: "projects", Raw: float64(p.ProjectsCount), Median: cfg.ProjectsMedian, Weight: cfg.ProjectsWeight},
		{Metric: "dailyAvg", Raw: p.DailyAvg, Median: cfg.DailyAvgMedian, Weight: cfg.DailyAvgWeight},
		{Metric: "hours", Raw: float64(p.TotalSeconds) / 3600, Median: cfg.HoursMedian, Weight: cfg.HoursWeight},
	}
	// First four use the exponential CDF (count-like, saturate fast); the two
	// volume metrics use log-normal (heavy-tailed) — the same split upstream
	// applies to commits/prs/issues/reviews vs stars/followers.
	var weightSum, blend float64
	for i := range subs {
		x := 0.0
		if subs[i].Median > 0 {
			x = subs[i].Raw / subs[i].Median
		}
		switch subs[i].Metric {
		case "dailyAvg", "hours":
			subs[i].Score = logNormalCDF(x)
		default:
			subs[i].Score = exponentialCDF(x)
		}
		blend += subs[i].Weight * subs[i].Score
		weightSum += subs[i].Weight
	}

	percentile := 100.0
	if weightSum > 0 {
		percentile = (1 - blend/weightSum) * 100
	}
	level := gradeLevels[len(gradeLevels)-1]
	for i, t := range gradeThresholds {
		if percentile <= t {
			level = gradeLevels[i]
			break
		}
	}
	return GradeResult{Level: level, Percentile: percentile, Subs: subs}
}

// longestStreak is the longest consecutive run of active days in the series.
func longestStreak(daily []int64) int {
	best, cur := 0, 0
	for _, s := range daily {
		if s > 0 {
			cur++
			best = max(best, cur)
		} else {
			cur = 0
		}
	}
	return best
}

File: web/src/features/publicprofile/grade.ts
// grade.ts — client-side port of internal/stats/grade.go (gaka-hsj port).
//
// Faithful port of github-readme-stats' rank algorithm: same two CDFs, same
// weighted blend, same percentile thresholds and level ladder. The metric
// set matches the Go version so the client-rendered badge on /p/:slug scores
// the same letter grade a widget SVG would.
//
// Kept as a small pure module (no React, no D3) so unit tests hit the
// algorithm directly. If the calibration ever drifts on the backend, add a
// port here and update tests in lockstep.
import type { PublicDashboardPayload } from "@/types/stats";

// Verbatim from internal/stats/grade.go: `gradeLevels` + `gradeThresholds`.
export const GRADE_LEVELS = ["S", "A+", "A", "A-", "B+", "B", "B-", "C+", "C"] as const;
const GRADE_THRESHOLDS = [1, 12.5, 25, 37.5, 50, 62.5, 75, 87.5, 100];

const exponentialCDF = (x: number) => 1 - Math.pow(2, -x);
const logNormalCDF = (x: number) => x / (1 + x);

// DefaultGradeConfig — mirrors internal/stats/grade.go DefaultGradeConfig.
// Kept as a plain export so future recalibration is a one-line diff.
export const DEFAULT_GRADE_CONFIG = {
  streakMedian: 5,
  streakWeight: 2,
  activeMedian: 50,
  activeWeight: 3,
  languagesMedian: 3,
  languagesWeight: 1,
  projectsMedian: 3,
  projectsWeight: 1,
  dailyAvgMedian: 7200, // seconds
  dailyAvgWeight: 4,
  hoursMedian: 40,
  hoursWeight: 1,
  minRangeDays: 7,
};

export interface GradeResult {
  level: (typeof GRADE_LEVELS)[number];
  percentile: number; // 0..100, LOWER is better
  subs: Array<{ metric: string; raw: number; median: number; score: number; weight: number }>;
}

function longestStreak(daily: number[]): number {
  let best = 0;
  let cur = 0;
  for (const s of daily) {
    if (s > 0) {
      cur++;
      if (cur > best) best = cur;
    } else {
      cur = 0;
    }
  }
  return best;
}

// Streak ENDING at "today" (the end of the range). Used by the
// current-streak-stat widget; not part of the grade blend.
export function currentStreak(daily: number[]): number {
  let cur = 0;
  for (let i = daily.length - 1; i >= 0; i--) {
    if (daily[i] > 0) cur++;
    else break;
  }
  return cur;
}

export function longestStreakInRange(daily: number[]): number {
  return longestStreak(daily);
}

// Grade takes a public dashboard payload and returns the same shape the
// backend's stats.Grade returns. `payload` is the minimum useful set of
// fields for the calculation — anything on PublicDashboardPayload works.
//
// The Go version uses `p.LanguagesCount` / `p.ProjectsCount` (true distinct
// counts). PublicDashboardPayload deliberately omits those (they leak
// hidden-value counts on short-list axes). The client-side port falls back
// to `p.languages.length` / `p.projects.length`, which under-reports for
// users with more than the top-N. Acceptable degradation: the top-N is
// deep enough (typically 8-10) that the calibration still produces the
// intended letter for most users, and the leaked-count avoidance is more
// important than a perfectly-aligned client badge.
export function computeGrade(
  payload: Pick<
    PublicDashboardPayload,
    "totalSeconds" | "dailyAvg" | "dailyTotal" | "languages" | "projects"
  >,
  cfg: typeof DEFAULT_GRADE_CONFIG = DEFAULT_GRADE_CONFIG,
): GradeResult {
  const rangeDays = payload.dailyTotal.length;
  let activeDays = 0;
  for (const s of payload.dailyTotal) if (s > 0) activeDays++;
  const denom = Math.max(rangeDays, cfg.minRangeDays);
  const activeRatio = (activeDays / denom) * 100;

  const subs = [
    { metric: "streak", raw: longestStreak(payload.dailyTotal), median: cfg.streakMedian, weight: cfg.streakWeight, score: 0 },
    { metric: "activeDays", raw: activeRatio, median: cfg.activeMedian, weight: cfg.activeWeight, score: 0 },
    { metric: "languages", raw: payload.languages.length, median: cfg.languagesMedian, weight: cfg.languagesWeight, score: 0 },
    { metric: "projects", raw: payload.projects.length, median: cfg.projectsMedian, weight: cfg.projectsWeight, score: 0 },
    { metric: "dailyAvg", raw: payload.dailyAvg, median: cfg.dailyAvgMedian, weight: cfg.dailyAvgWeight, score: 0 },
    { metric: "hours", raw: payload.totalSeconds / 3600, median: cfg.hoursMedian, weight: cfg.hoursWeight, score: 0 },
  ];

  let weightSum = 0;
  let blend = 0;
  for (const s of subs) {
    const x = s.median > 0 ? s.raw / s.median : 0;
    s.score = s.metric === "dailyAvg" || s.metric === "hours" ? logNormalCDF(x) : exponentialCDF(x);
    blend += s.weight * s.score;
    weightSum += s.weight;
  }
  const percentile = weightSum > 0 ? (1 - blend / weightSum) * 100 : 100;
  let level: (typeof GRADE_LEVELS)[number] = GRADE_LEVELS[GRADE_LEVELS.length - 1];
  for (let i = 0; i < GRADE_THRESHOLDS.length; i++) {
    if (percentile <= GRADE_THRESHOLDS[i]) {
      level = GRADE_LEVELS[i];
      break;
    }
  }
  return { level, percentile, subs };
}
